### DISCUSSION TITLE
Start time generation, instead of build time

### DISCUSSION THREAD

**Discussion by raphaeljoie** [vercel/next.js#24046] [Ideas]
### Describe the feature you'd like to request

When delivering containerised solutions, I'm used to package built version of the services. So do I with NextJS. 

Between two deployments, the only difference is a `BACKEND_HOST` variable, to be provided in the ENV.
```shell
docker run -e BACKEND_HOST=http://start.sh nextjs-fancy-app
```
From the doc, I read that the way to expose runtime config to the client side is using `next.config.js`
```js
// next.config.js
module.exports = {
  publicRuntimeConfig: {
    backendHost: process.env['BACKEND_HOST'],
  },
}
```
later I have a js file to keep the backend API
```js
// api.js
import getConfig from 'next/config'
const{ publicRuntimeConfig } = getConfig()

export default const api = {
  create: () => fetch(publicRuntimeConfig.backendHost /*...*/)
}
```
and
```sh
BACKEND_HOST=http://build.sh npm run build
BACKEND_HOST=http://start.sh npm run start
```
Problem: Static and SSG pages are using `http://build.sh`, as expected, because the environment is being freezed by the static and SSG process at build time. The only way I found to make it work, is to use SSR. Which is quite sad, just for one single environment variable that is not changing between renderings

### Describe the solution you'd like

What about an "BSSG" (Boot server side generation), working like ISR but only once before first rendering?

```js
// /pages/index.js

import getConfig from 'next/config'
const{ publicRuntimeConfig } = getConfig()

export async function getStaticProps(context) {
  return {
    props: {backendHost: publicRuntimeConfig.backendHost},
    start: true
  }
}
```

### Describe alternatives you've considered

I have considered `ISR`
* with `0` value but it is not accepted
* with `1` as a value, but the time is being counted after the first render => first render is NOK

Now the only solution I found is to enabled SSR for all the pages :'( 

PS: the other alternative is of course
```json
// package.json
{
  "scripts": {
    "dev": "next dev -p \"${PORT:-80}\"",
    "build": "next build",
    "start": "next build && next start -p \"${PORT:-80}\""
  },
```
but it is not very efficient as
* build at boot time doesn't make sense
* it means that the `node_modules` directory with the dependencies must be shipped with the container

**Comment by djsilcock** [comment]
Bit of a fudge admittedly, but you could set up an API route or even a static file which served the config you needed, import that in your _app.js and serve through context? 
Think what is needed is some variant of #10949
